routing/http/server/server_test.go

```diff
@@ -174,29 +188,51 @@ func TestProviders(t *testing.T) {
+	})
+
+	t.Run("JSON Response with addr filtering", func(t *testing.T) {
+		runTest(t, mediaTypeJSON, "webtransport,!p2p-circuit", "", false, false, `{"Providers":[{"Addrs":["/ip4/8.8.8.8/udp/4001/quic-v1/webtransport"],"ID":"12D3KooWM8sovaEGU1bmiWGWAzvs47DEcXKZZTuJnpQyVTkRs2Vn","Protocols":["transport-bitswap"],"Schema":"peer"}]}`)
+	})
+
+	t.Run("JSON Response with protocol and addr filtering", func(t *testing.T) {
+		runTest(t, mediaTypeJSON, "quic-v1", "transport-bitswap", false, false,
+			`{"Providers":[{"Addrs":["/ip4/127.0.0.1/udp/4001/quic-v1","/ip4/102.101.1.1/udp/4001/quic-v1/webtransport/p2p/12D3KooWEjsGPUQJ4Ej3d1Jcg4VckWhFbhc6mkGunMm1faeSzZMu/p2p-circuit","/ip4/8.8.8.8/udp/4001/quic-v1/webtransport"],"ID":"12D3KooWM8sovaEGU1bmiWGWAzvs47DEcXKZZTuJnpQyVTkRs2Vn","Protocols":["transport-bitswap"],"Schema":"peer"}]}`)
+	})
+
+	t.Run("JSON Response with protocol filtering", func(t *testing.T) {
+		runTest(t, mediaTypeJSON, "", "transport-ipfs-gateway-http", false, false,
+			`{"Providers":[{"Addrs":[],"ID":"12D3KooWM8sovaEGU1bmiWGWAzvs47DEcXKZZTuJnpQyVTkRs2Vz","Protocols":["transport-ipfs-gateway-http"],"Schema":"peer"}]}`)
 	})
 
 	t.Run("Empty JSON Response", func(t *testing.T) {
-		runTest(t, mediaTypeJSON, true, false, `{"Providers":null}`)
+		runTest(t, mediaTypeJSON, "", "", true, false, `{"Providers":null}`)
 	})
 
 	t.Run("Wildcard Accept header defaults to JSON Response", func(t *testing.T) {
 		accept := "text/html,*/*"
-		runTest(t, accept, true, false, `{"Providers":null}`)
+		runTest(t, accept, "", "", true, false, `{"Providers":null}`)
 	})
 
 	t.Run("Missing Accept header defaults to JSON Response", func(t *testing.T) {
 		accept := ""
-		runTest(t, accept, true, false, `{"Providers":null}`)
+		runTest(t, accept, "", "", true, false, `{"Providers":null}`)
 	})
 
 	t.Run("NDJSON Response", func(t *testing.T) {
-		runTest(t, mediaTypeNDJSON, false, true, `{"Addrs":[],"ID":"12D3KooWM8sovaEGU1bmiWGWAzvs47DEcXKZZTuJnpQyVTkRs2Vn","Protocols":["transport-bitswap"],"Schema":"peer"}`+"\n"+`{"Schema":"bitswap","Protocol":"transport-bitswap","ID":"12D3KooWM8sovaEGU1bmiWGWAzvs47DEcXKZZTuJnpQyVTkRs2Vz"}`+"\n")
+		runTest(t, mediaTypeNDJSON, "", "", false, true, `{"Addrs":["/ip4/127.0.0.1/tcp/4001","/ip4/127.0.0.1/udp/4001/quic-v1","/ip4/127.0.0.1/tcp/4001/ws","/ip4/102.101.1.1/tcp/4001/p2p/12D3KooWEjsGPUQJ4Ej3d1Jcg4VckWhFbhc6mkGunMm1faeSzZMu/p2p-circuit","/ip4/102.101.1.1/udp/4001/quic-v1/webtransport/p2p/12D3KooWEjsGPUQJ4Ej3d1Jcg4VckWhFbhc6mkGunMm1faeSzZMu/p2p-circuit","/ip4/8.8.8.8/udp/4001/quic-v1/webtransport"],"ID":"12D3KooWM8sovaEGU1bmiWGWAzvs47DEcXKZZTuJnpQyVTkRs2Vn","Protocols":["transport-bitswap"],"Schema":"peer"}`+"\n"+`{"Schema":"bitswap","Protocol":"transport-bitswap","ID":"12D3KooWM8sovaEGU1bmiWGWAzvs47DEcXKZZTuJnpQyVTkRs2Vz"}`+"\n")
```

Comment:
@2color CI fails on this equals test ([log](https://github.com/ipfs/boxo/actions/runs/10834542823/job/30063981429?pr=671#step:13:324)), perhaps because  iteration order on maps is not specified in golang, so these tests worked just by luck, and once we added more logic, internal hashes changed and the order changed? 

Potential fix is to refactor `runTest` to do normalize json before comparing

Reply:
Actually the tests were wrong (I forgot to update the assertions after changing the test data. This was not a normalisation problem. 

The last commits should fix this.